src/expr/src/explain/text.rs

```diff
@@ -101,6 +104,48 @@ where
             for (id, op) in self.sources.iter().filter(|(_, op)| !op.is_identity()) {
                 writeln!(f, "{}Source {}", ctx.indent, id)?;
                 ctx.indented(|ctx| op.fmt_text(f, ctx))?;
+                if self.context.config.mfp_pushdown {
```

Comment:
This type of analysis should not happen in the `DisplayText` implementation, but rather in the various `Explain` implementations (use your IDE to find them).

Basically `Explain` takes an explainee type (for example `MirRelationExpr` or `DataflowDescription<MirRelationExpr>`, and then it does two things:

1. Pre-processing (tree normalization, attribute derivation).
2. Information bundling - the original input, together with all derived information needs to be bundled into a structure that 
    1. implements `DisplayText` (in the `explain_text` implementation), or
    2. implements `DisplayJson` (in the `explain_json` implementation).

At the moment we use a finite set of such structures:

1. `ExplainSinglePlan` for explain outputs that consist of single plan (used for `EXPLAIN RAW` and `EXPLAIN DECORRELATED`).
2. `ExplainMultiPlan` for explain outputs that consist of multiple plan and optional source information.

I think the best thing to do is to tweak the various `as_explain_single_plan` methods used to implement the `Explain` interface for:

1. `DataflowDescription<Plan>`,
2. `DataflowDescription<OptimizedMirRelationExpr>`, and 
3. `Explainable<'a, DataflowDescription<OptimizedMirRelationExpr>>`

So the `sources` bound to the `ExplainMultiPlan::sources` field are pre-processed with the `MfpPushdown` transform. 

Feel free to tweak `sources` or add a new field next to `sources` that represents the `MfpPushdown` results of the `sources` vector.

In the long term, you can pass the `PersistSourceDataStats` implementation through an additional `ExplainContext` field, but this can be a follow-up PR.

Reply:
Hey, thanks for the feedback!

> I think the best thing to do is to tweak the various `as_explain_single_plan` methods used to implement the Explain interface for [...]

Based on the code locations you linked I'm reasonably sure you meant `as_explain_multi_plan` here, since that matches the three examples you gave and also is the place at which source info appears. (Let me know if I have assumed incorrectly!)

I've updated the branch with a new `ExplainMultiPlanSource` struct which contains some optional pushdown info. I've attempted to stick to the existing style as much as possible, but let me know if you'd like to see any changes there.

Agree there may be things that can be factored out directly into the config in the future... I think we'll need to iterate a bit more on our plans before it's worth making the switch, but I'll make a note to follow up once things have stabilized.

Reply:
Thanks, @bkirwi. Can you re-request a review from me from the UI? I will check the changes first thing on Monday.